Code/GraphMol/Descriptors/Property.h

```diff
@@ -0,0 +1,242 @@
+//       products derived from this software without specific prior written
+//       permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+#ifndef RDKIT_PROPERTIES_H
+#define RDKIT_PROPERTIES_H
+
+#include <GraphMol/RDKitBase.h>
+#include <string>
+#include <boost/shared_ptr.hpp>
+
+namespace RDKit {
+namespace Descriptors {
+  
+// Computes an additive property from a molecule
+struct PropertyFxn {
+  virtual ~PropertyFxn() {}
+  
+  //! Compute the value for the molecule
+  virtual double compute(const RDKit::ROMol &mol) const = 0;
+  
+  //! Is the property additive with respect to molecular fragments?
+  virtual bool isAdditive() const { return false; }
+};
+
+//! Defines an additive property used to filter molecules
+/* Known properties include
+   MW - Molecular weight
+   TPSA - polar surface area
+   ALOGP - atom based log P
+   NumRotors - the number of rotors (including side chain bond)
+   MissingStereo - the number of atoms with unassigned stereo
+   LipinskiHBA
+   LipinskiHBD
+   HBA
+   HBD
+   NumHeteroAtoms
+   NumAmideBonds
+   FractionCSP3
+   NumRings
+   NumAromaticRings
+   NumAliphaticRings
+   NumSaturatedRings
+   NumHeteroCycles
+   NumSaturatedHeteroCycles
+   NumAromaticHeteroCycles
+   NumAliphaticCarboCycles
+   NumAromaticCarboCycles
+   NumSaturatedCarboCycles
+   NumAlpiphaticHeteroCycles
+   NumSpiroAtoms
+   NumBridgeheadAtoms
+   AMW
+   LabuteASA
+
+   user defined types can be added as well with an appropriate
+    computing functor.
+
+   Example:
+   \verbatim
+     Property mw(Property::MW);
+     Property logp(Property::ALOGP);
+
+     // User based function
+     struct MyPropertyFunction : public PropertyFxn {
+       double compute(const ROMol &mol) const {
+         return rdcast<double>(mol.getNumAtoms());
+       }
+       bool isAdditive() const { return true; }
+     }
+
+     // Make the a num atom property
+     boost::shared_ptr<PropertyFxn> atomfxn(new MyPropertyFxn);
+     Property numAtoms("NumAtoms", atomfxn);
+
+     double alogp = logp.computeProperty(mol);
+     double nAtoms = numAtoms.computeProperty(mol);
+     
+    \endverbatim
+
+*/
+
+class Property {
+public:  
+  enum PropType{ MW=0,
```

Comment:
Not particularly except that they avoid a small class of bugs with enums.  I had wanted to use an enum to be very explicit in what the calculable properties are, but you are correct that this is yet another place that needs an update when new props are added.

We could use a registration system, i.e.:
```
double myNewProp(ROMol &mol) {
}

// function, short name, description, version
REGISTER_PROPERTY( myNewProp, "MyNewProp", "My Fabulous new Property", "1.0.0" );
```

And get rid of the enums at the expense of some explicitness.